presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java

```diff
@@ -0,0 +1,229 @@
+        {
+            ImmutableMap.Builder<Symbol, Signature> functionsBuilder = ImmutableMap.builder();
+            ImmutableList.Builder<Symbol> groupByKeys = ImmutableList.builder();
+
+            List<Symbol> intersectOutputSymbols = node.getOutputSymbols();
+            Signature countSignature = new Signature("count", AGGREGATE, "bigint", "boolean");
+
+            for (int i = intersectOutputSymbols.size(); i < outputSymbols.size(); i++) {
+                Symbol countSymbol = allocateNewSymbolFrom("count", BIGINT);
+                aggregationsBuilder.put(countSymbol, new FunctionCall(QualifiedName.of("count"), ImmutableList.<Expression>of(new QualifiedNameReference(outputSymbols.get(i).toQualifiedName()))));
+                functionsBuilder.put(countSymbol, countSignature);
+            }
+
+            for (Symbol nodeSymbol : intersectOutputSymbols) {
+                int fieldIndex = intersectOutputSymbols.indexOf(nodeSymbol);
+                Symbol outputSymbol = outputSymbols.get(fieldIndex);
+                groupByKeys.add(outputSymbol);
+                outputProjections.put(nodeSymbol, new QualifiedNameReference(outputSymbol.toQualifiedName()));
+            }
+
+            return new AggregationNode(idAllocator.getNextId(),
+                    sourceNode,
+                    groupByKeys.build(),
+                    aggregationsBuilder.build(),
+                    functionsBuilder.build(),
+                    ImmutableMap.<Symbol, Symbol>of(),
+                    Step.SINGLE,
+                    Optional.empty(),
+                    1.0,
+                    Optional.empty());
+        }
+
+        private Symbol allocateNewSymbolFrom(String name, Type type)
+        {
+            return symbolAllocator.newSymbol(name, type);
+        }
+
+        private ProjectNode createProjectNode(IntersectNode intersectNode, List<PlanNode> sources, int sourceIndex)
+        {
+            ImmutableMap.Builder<Symbol, Expression> projectionAssignments = ImmutableMap.builder();
+            Map<Symbol, QualifiedNameReference> sourceSymbolMap = intersectNode.sourceSymbolMap(sourceIndex);
+
+            // add existing intersect symbols to projection
+            for (Symbol intersectSymbol : sourceSymbolMap.keySet()) {
+                Symbol symbol = allocateNewSymbolFrom(intersectSymbol);
+                projectionAssignments.put(symbol, sourceSymbolMap.get(intersectSymbol));
+            }
+
+            // add extra marker fields to the projection
+            List<Expression> markers = new ArrayList<>(Collections.nCopies(sources.size(), new Cast(new NullLiteral(), StandardTypes.BOOLEAN)));
+            markers.set(sourceIndex, TRUE_LITERAL);
+            for (Expression markerExpression : markers) {
+                projectionAssignments.put(allocateNewSymbolFrom(INTERSECT_MARKER, BOOLEAN), markerExpression);
```

Comment:
What do you mean by source? I can add the `id` of the source node to the symbol name. But `EXPLAIN` does not show the id of the nodes, so  it wont be obvious which source the marker belongs to.